File: ios/LingoPlay/NeuralTextToSpeech.swift
import AVFoundation
import Foundation
@preconcurrency import SherpaOnnx

enum TTSRoute: Equatable {
    case system
    case neural
}

enum TTSRoutingPolicy {
    static func route(
        targetLanguage: String,
        preferredVoiceIdentifier: String?,
        neuralVoiceInstalled: Bool
    ) -> TTSRoute {
        let baseLanguage = targetLanguage.lowercased().split(separator: "-").first.map(String.init)
        return baseLanguage == "vi" &&
            preferredVoiceIdentifier == NeuralVoicePackManifest.voiceIdentifier &&
            neuralVoiceInstalled
            ? .neural
            : .system
    }
}

enum NeuralTTSPerformancePolicy {
    static func threadCount(availableProcessors: Int) -> Int {
        min(2, max(1, max(1, availableProcessors) / 2))
    }
}

actor NeuralVietnameseTTSService {
    func synthesize(
        document: TranslationDocument,
        model: InstalledNeuralVoice,
        progress: @MainActor @Sendable (Int, Int) -> Void
    ) async throws -> DubSpeechDocument {
        let baseLanguage = document.targetLanguage.lowercased().split(separator: "-").first.map(String.init)
        guard baseLanguage == "vi" else {
            throw TTSError.synthesisFailed("The installed Neural Voice supports Vietnamese output only.")
        }

        let vits = sherpaOnnxOfflineTtsVitsModelConfig(
            model: model.modelURL.path,
            lexicon: "",
            tokens: model.tokensURL.path,
            dataDir: model.dataDirectoryURL.path
        )
        let modelConfig = sherpaOnnxOfflineTtsModelConfig(
            vits: vits,
            numThreads: NeuralTTSPerformancePolicy.threadCount(
                availableProcessors: ProcessInfo.processInfo.activeProcessorCount
            ),
            provider: "cpu"
        )
        var config = sherpaOnnxOfflineTtsConfig(model: modelConfig, silenceScale: 0.2)
        let tts = SherpaOnnxOfflineTtsWrapper(config: &config)
        guard tts.tts != nil else {
            throw TTSError.synthesisFailed("Unable to initialize the installed Neural Voice model.")
        }

        let root = try makeSessionDirectory()
        var succeeded = false
        defer {
            if !succeeded { try? FileManager.default.removeItem(at: root) }
        }
        var output: [DubSpeechSegment] = []
        for (index, segment) in document.segments.enumerated() {
            output.append(try synthesizeSegment(segment, tts: tts, root: root))
            await progress(index + 1, document.segments.count)
        }
        succeeded = true
        return DubSpeechDocument(
            voiceIdentifier: NeuralVoicePackManifest.voiceIdentifier,
            segments: output
        )
    }

    private func synthesizeSegment(
        _ segment: TranslationSegment,
        tts: SherpaOnnxOfflineTtsWrapper,
        root: URL
    ) throws -> DubSpeechSegment {
        let targetMs = DurationFitPolicy.targetDurationMs(startMs: segment.startMs, endMs: segment.endMs)
        var multiplier: Float = 1

        for attempt in 0..<DurationFitPolicy.maximumAttempts {
            let fileURL = root.appendingPathComponent("\(segment.id)-\(attempt).wav")
            try? FileManager.default.removeItem(at: fileURL)
            let audio = tts.generate(text: segment.translatedText, sid: 0, speed: multiplier)
            guard audio.audio != nil, audio.n > 0, audio.sampleRate > 0 else {
                throw TTSError.synthesisFailed("Neural Voice produced no audio for segment \(segment.id).")
            }
            guard audio.save(filename: fileURL.path) == 1,
                  FileManager.default.fileExists(atPath: fileURL.path)
            else {
                throw TTSError.synthesisFailed("Neural Voice could not save segment \(segment.id).")
            }

            let durationMs = max(
                1,
                Int((Double(audio.n) * 1_000 / Double(audio.sampleRate)).rounded())
            )
            let fits = DurationFitPolicy.fits(actualMs: durationMs, targetMs: targetMs)
            let next = DurationFitPolicy.nextRateMultiplier(
                actualMs: durationMs,
                targetMs: targetMs,
                current: multiplier
            )
            let finalAttempt = attempt == DurationFitPolicy.maximumAttempts - 1
            if fits || next == nil || finalAttempt {
                return DubSpeechSegment(
                    id: segment.id,
                    startMs: segment.startMs,
                    endMs: DurationFitPolicy.effectiveEndMs(
                        startMs: segment.startMs,
                        sourceEndMs: segment.endMs,
                        speechDurationMs: durationMs
                    ),
                    audioURL: fileURL,
                    speechDurationMs: durationMs,
                    tailSilenceMs: DurationFitPolicy.tailSilenceMs(
                        actualMs: durationMs,
                        targetMs: targetMs
                    ),
                    rateMultiplier: multiplier
                )
            }
            try? FileManager.default.removeItem(at: fileURL)
            multiplier = next ?? multiplier
        }
        throw TTSError.synthesisFailed("No Neural Voice synthesis attempt was executed.")
    }

    private func makeSessionDirectory() throws -> URL {
        let caches = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
        let directory = caches
            .appendingPathComponent("LingoPlay/NeuralTTS", isDirectory: true)
            .appendingPathComponent(UUID().uuidString, isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory
    }
}

@MainActor
final class OfflineDubbingTTSService {
    private let system = SystemVietnameseTTSService()
    private let neural = NeuralVietnameseTTSService()
    private let neuralStore = NeuralVoiceModelStore()
    private let cloning = VoiceCloningTTSService()
    private let cloningStore = VoiceCloningModelStore()

    func synthesize(
        document: TranslationDocument,
        preferredVoiceIdentifier: String?,
        speakerVoiceMap: [String: String] = [:],
        cloneReferences: [String: VoiceCloneReference] = [:],
        progress: @MainActor @Sendable (Int, Int) -> Void
    ) async throws -> DubSpeechDocument {
        guard !cloneReferences.isEmpty else {
            return try await synthesizeRegular(
                document: document,
                preferredVoiceIdentifier: preferredVoiceIdentifier,
                speakerVoiceMap: speakerVoiceMap,
                progress: progress
            )
        }

        let cloneable = document.segments.filter { segment in
            segment.overlappingSpeakerIDs.isEmpty &&
                segment.speakerID.flatMap { cloneReferences[$0] } != nil
        }
        let cloneableIDs = Set(cloneable.map(\.id))
        let fallback = document.segments.filter { !cloneableIDs.contains($0.id) }
        var output: [DubSpeechSegment] = []
        var completed = 0

        if !cloneable.isEmpty {
            guard let model = cloningStore.model() else {
                throw TTSError.synthesisFailed("Voice Cloning model is not installed.")
            }
            let cloned = try await cloning.synthesize(
                document: TranslationDocument(
                    sourceLanguage: document.sourceLanguage,
                    targetLanguage: document.targetLanguage,
                    segments: cloneable,
                    mode: document.mode,
                    speakerVoiceMap: document.speakerVoiceMap
                ),
                model: model,
                references: cloneReferences
            ) { local, _ in
                progress(completed + local, document.segments.count)
            }
            output.append(contentsOf: cloned.segments)
            completed += cloneable.count
        }

        if !fallback.isEmpty {
            let normal = try await synthesizeRegular(
                document: TranslationDocument(
                    sourceLanguage: document.sourceLanguage,
                    targetLanguage: document.targetLanguage,
                    segments: fallback,
                    mode: document.mode,
                    speakerVoiceMap: document.speakerVoiceMap
                ),
                preferredVoiceIdentifier: preferredVoiceIdentifier,
                speakerVoiceMap: speakerVoiceMap
            ) { local, _ in
                progress(completed + local, document.segments.count)
            }
            output.append(contentsOf: normal.segments)
        }

        let order = Dictionary(uniqueKeysWithValues: document.segments.enumerated().map { ($1.id, $0) })
        output.sort { (order[$0.id] ?? .max) < (order[$1.id] ?? .max) }
        return DubSpeechDocument(voiceIdentifier: "hybrid:clone-local", segments: output)
    }

    private func synthesizeRegular(
        document: TranslationDocument,
        preferredVoiceIdentifier: String?,
        speakerVoiceMap: [String: String],
        progress: @MainActor @Sendable (Int, Int) -> Void
    ) async throws -> DubSpeechDocument {
        if speakerVoiceMap.isEmpty {
            return try await synthesizeWithVoice(
                document: document,
                preferredVoiceIdentifier: preferredVoiceIdentifier,
                progress: progress
            )
        }

        var grouped: [(voiceID: String?, segments: [TranslationSegment])] = []
        for segment in document.segments {
            let voiceID = segment.speakerID.flatMap { speakerVoiceMap[$0] } ?? preferredVoiceIdentifier
            if let index = grouped.firstIndex(where: { $0.voiceID == voiceID }) {
                grouped[index].segments.append(segment)
            } else {
                grouped.append((voiceID, [segment]))
            }
        }

        var output: [DubSpeechSegment] = []
        var completed = 0
        for group in grouped {
            let partial = try await synthesizeWithVoice(
                document: TranslationDocument(
                    sourceLanguage: document.sourceLanguage,
                    targetLanguage: document.targetLanguage,
                    segments: group.segments,
                    mode: document.mode,
                    speakerVoiceMap: document.speakerVoiceMap
                ),
                preferredVoiceIdentifier: group.voiceID
            ) { local, _ in
                progress(completed + local, document.segments.count)
            }
            output.append(contentsOf: partial.segments)
            completed += group.segments.count
        }
        let order = Dictionary(uniqueKeysWithValues: document.segments.enumerated().map { ($1.id, $0) })
        output.sort { (order[$0.id] ?? .max) < (order[$1.id] ?? .max) }
        return DubSpeechDocument(
            voiceIdentifier: "multi-speaker:\(grouped.count)",
            segments: output
        )
    }

    private func synthesizeWithVoice(
        document: TranslationDocument,
        preferredVoiceIdentifier: String?,
        progress: @MainActor @Sendable (Int, Int) -> Void
    ) async throws -> DubSpeechDocument {
        let neuralModel = neuralStore.model()
        switch TTSRoutingPolicy.route(
            targetLanguage: document.targetLanguage,
            preferredVoiceIdentifier: preferredVoiceIdentifier,
            neuralVoiceInstalled: neuralModel != nil
        ) {
        case .neural:
            return try await neural.synthesize(
                document: document,
                model: neuralModel!,
                progress: progress
            )
        case .system:
            let systemIdentifier = preferredVoiceIdentifier == NeuralVoicePackManifest.voiceIdentifier
                ? nil
                : preferredVoiceIdentifier
            return try await system.synthesize(
                document: document,
                preferredVoiceIdentifier: systemIdentifier,
                progress: progress
            )
        }
    }
}

File: android/app/src/main/java/com/lingoplay/app/NeuralTextToSpeech.kt
package com.lingoplay.app

import android.content.Context
import com.k2fsa.sherpa.onnx.OfflineTts
import com.k2fsa.sherpa.onnx.OfflineTtsConfig
import com.k2fsa.sherpa.onnx.OfflineTtsModelConfig
import com.k2fsa.sherpa.onnx.OfflineTtsVitsModelConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.util.UUID
import kotlin.math.max
import kotlin.math.roundToInt

internal enum class TTSRoute {
    SYSTEM,
    NEURAL,
}

internal object TTSRoutingPolicy {
    fun route(
        targetLanguage: String,
        preferredVoiceId: String?,
        neuralVoiceInstalled: Boolean,
    ): TTSRoute =
        if (
            targetLanguage.substringBefore('-').equals("vi", ignoreCase = true) &&
            preferredVoiceId == NeuralVoicePackManifest.voiceId &&
            neuralVoiceInstalled
        ) {
            TTSRoute.NEURAL
        } else {
            TTSRoute.SYSTEM
        }
}

object OfflineDubbingTTSService {
    suspend fun availableVoices(context: Context): List<OfflineVoiceOption> {
        val system = runCatching {
            SystemVietnameseTTSService.availableOfflineVoices(context)
        }.getOrDefault(emptyList())
        return (listOfNotNull(NeuralVoiceModelStore.voiceOption(context)) + system)
            .distinctBy(OfflineVoiceOption::id)
    }

    suspend fun synthesize(
        context: Context,
        document: TranslationDocument,
        preferredVoiceId: String?,
        onProgress: suspend (segment: Int, total: Int) -> Unit,
    ): DubSpeechDocument {
        val neuralModel = NeuralVoiceModelStore.find(context)
        return when (
            TTSRoutingPolicy.route(
                targetLanguage = document.targetLanguage,
                preferredVoiceId = preferredVoiceId,
                neuralVoiceInstalled = neuralModel != null,
            )
        ) {
            TTSRoute.NEURAL -> NeuralVietnameseTTSService.synthesize(
                context = context,
                document = document,
                model = checkNotNull(neuralModel),
                onProgress = onProgress,
            )
            TTSRoute.SYSTEM -> SystemVietnameseTTSService.synthesize(
                context = context,
                document = document,
                preferredVoiceName = preferredVoiceId?.takeUnless { it == NeuralVoicePackManifest.voiceId },
                onProgress = onProgress,
            )
        }
    }
}

object NeuralVietnameseTTSService {
    suspend fun synthesize(
        context: Context,
        document: TranslationDocument,
        model: NeuralVoiceModel,
        onProgress: suspend (segment: Int, total: Int) -> Unit,
    ): DubSpeechDocument = withContext(Dispatchers.Default) {
        require(document.targetLanguage.substringBefore('-').equals("vi", ignoreCase = true)) {
            "The installed Neural Voice supports Vietnamese output only."
        }

        val vits = OfflineTtsVitsModelConfig(
            model = model.model.absolutePath,
            tokens = model.tokens.absolutePath,
            dataDir = model.dataDir.absolutePath,
        )
        val modelConfig = OfflineTtsModelConfig(
            vits = vits,
            numThreads = NeuralTTSPerformancePolicy.threadCount(Runtime.getRuntime().availableProcessors()),
            provider = "cpu",
        )
        val tts = OfflineTts(
            assetManager = null,
            config = OfflineTtsConfig(model = modelConfig, silenceScale = 0.2f),
        )
        val root = File(context.cacheDir, "lingoplay/neural-tts/${UUID.randomUUID()}").apply { mkdirs() }
        var succeeded = false
        try {
            val output = mutableListOf<DubSpeechSegment>()
            document.segments.forEachIndexed { index, segment ->
                output += synthesizeSegment(tts, segment, root)
                onProgress(index + 1, document.segments.size)
            }
            succeeded = true
            DubSpeechDocument(NeuralVoicePackManifest.voiceId, output)
        } finally {
            tts.release()
            if (!succeeded) root.deleteRecursively()
        }
    }

    private fun synthesizeSegment(
        tts: OfflineTts,
        segment: TranslationSegment,
        root: File,
    ): DubSpeechSegment {
        val targetMs = DurationFitPolicy.targetDurationMs(segment.startMs, segment.endMs)
        var multiplier = 1.0f

        repeat(DurationFitPolicy.MAXIMUM_ATTEMPTS) { attempt ->
            val output = File(root, "${segment.id}-$attempt.wav")
            output.delete()
            val audio = tts.generate(segment.translatedText, sid = 0, speed = multiplier)
            check(audio.samples.isNotEmpty() && audio.sampleRate > 0) {
                "Neural Voice produced no audio for segment ${segment.id}."
            }
            check(audio.save(output.absolutePath) && output.isFile && output.length() > 44L) {
                "Neural Voice could not save segment ${segment.id}."
            }
            val durationMs = max(
                1,
                (audio.samples.size.toDouble() * 1_000.0 / audio.sampleRate.toDouble()).roundToInt(),
            )
            val fits = DurationFitPolicy.fits(durationMs, targetMs)
            val next = DurationFitPolicy.nextRateMultiplier(durationMs, targetMs, multiplier)
            val finalAttempt = attempt == DurationFitPolicy.MAXIMUM_ATTEMPTS - 1
            if (fits || next == null || finalAttempt) {
                return DubSpeechSegment(
                    id = segment.id,
                    startMs = segment.startMs,
                    endMs = DurationFitPolicy.effectiveEndMs(segment.startMs, segment.endMs, durationMs),
                    audioFile = output,
                    speechDurationMs = durationMs,
                    tailSilenceMs = DurationFitPolicy.tailSilenceMs(durationMs, targetMs),
                    rateMultiplier = multiplier,
                )
            }
            output.delete()
            multiplier = next
        }
        error("No Neural Voice synthesis attempt was executed.")
    }
}

internal object NeuralTTSPerformancePolicy {
    fun threadCount(availableProcessors: Int): Int =
        (availableProcessors.coerceAtLeast(1) / 2).coerceIn(1, 2)
}
